### ISSUE TITLE
Setting auto reply message once creates unwanted system overhead past message end date

### ISSUE THREAD

**Issue by ariabamdad** [postfixadmin/postfixadmin#832]
Hi,
Currently if a user configures a vacation auto response message and turns it on once, then postfixadmin will update the alias table’s goto field with the normal alias as well as the properly formatted user#example.com@autoreply.example.com address.  This autoreply address remains in the ‘goto’ field regardless of the setting of the ‘active’ field in the vacation table after the message end data has passed.  The only time the autoreply alias is removed from the goto field of the alias table is when the user manually reviews the auto reply form and then clicks Remove away message.  Only then will the goto field get updated and the autoreply address removed.

This adds unnecessary overhead to a system because every user at some point will set an auto reply message but from that point forward, the mail system is forced to call the vacation.pl script for every message the user will receive, regardless of the setting of the ‘active’ field in the vacation table.  This is unnecessary and we can’t expect users to return and click ‘remove away message’.

The proper solution is to have the vacation.pl script automatically update and remove the added autoreply e-mail address from the goto field of the alias table the next time vacation.pl is triggered for a message for a user that either has active set to false in the vacation table or the current date is past the end data of the auto reply message. The logic for this is the same as what vacation.pl currently has in order to decide if a message is to be sent or not.  What is missing is the code to update the alias field.  My lack of experience with php prevents me to contribute here.
 
The suggestion in #616  “Expired vacation entries don't get deactivated” to update the vacation table via a cron script will not address the problem above.

Is there a similar call to a function/method that can be added to the suggested solution above that could also update the goto field of the alias table, removing the autoreply address from the field, until the vacation script addresses this problem?

Thanks,
Aria

**Comment by ariabamdad** [comment]
Hi David.  Thank you for your prompt reply.

> The easiest might be some sort of cron job which looks for active vacation(s) and if they need deactivating it removes them?

Yes, that is what you suggested here #616 but the problem there is that simply setting active to 0 for the vacation table does not have any impact on the fact that the alias still references the autoreply domain, causing postfix to attempt delivery to both the mailbox and the autoreply domain.  That's what I am trying to avoid because over time, nearly every user will set an autoreply message and once set, every message addressed to them will trigger the vacation script forever!  

> 
> Alternatively, I suppose the alias lookup SQL could be changed so it adds in the vacation alias if the time window is correct? I'm not sure if Postfix would be happy receiving two rows from doing some sort of UNION ALL on the vacation table or not though ....
> 
> ```
> MariaDB [postfixadmin]> select goto from alias WHERE address = 'test@example.com' UNION ALL ( Select concat(replace(email, '@', '#'), '@autoreply.mydomain.com')  as goto FROM vacation WHERE email = 'test@example.com' AND activ
> euntil >= NOW() AND activefrom <= NOW());
> +-----------------------------------------+
> | goto                                    |
> +-----------------------------------------+
> | test@example.com                        |
> | test#example.com@autoreply.mydomain.com |
> +-----------------------------------------+
> 2 rows in set (0.001 sec)
> ```

Oddly enough, the above query returns a single row with the two e-mails together for me.  Not sure why.  Regardless, as you said, I am not sure it postfix cares it it gets one row or two and the problem remains.  I also prefer not to hard code the autoreply domain here because that is a configuration option for PFA.

In my case, the user is not allowed to change their alias so all of my alias table address fields match the goto field.  Since this is the case, I can combine in one cron the search for vacation settings that have expired, turn them to active=0 and then replace the alias table goto field for that email (where email=address) with the value in the address field.  This will remove the autoreply address from the goto field and in my unique case, returns the goto field to the default (matching the address field).  

The ideal solution would be to perform the same database operation the GUI performs which is to remove the autoreply address from the set of addresses in the goto field of the alias table but I am not sure how to implement that outside of the GUI php code.

Thank you again.

**Comment by ariabamdad** [comment]
Super.  Thank you David.  I will test this code and let you know if I find any issues.  Much appreciated.

**Comment by ariabamdad** [comment]
@DavidGoodwin thank you again for your code above.  I have tested it and it works as a standalone module that I will call via cron.  Eventually I think this code should be part of the vacation.pl script so that when the script is triggered for a user that has an expired auto reply interval, the same functions are called to clear everything.

In my testing, I found a small bug in the vacation.pl script with respect to logging.   The last line of code at the end of the **sub send_vacation_email** subroutine that logs a statement when a message is sent should be $logger->info  but instead it is a $logger->debug.  This will result in no messages being logged to the log when a auto reply is sent.  The only time message is logged is if the auto reply interval has elapsed.

Thank you again,
Aria
